src/reporting/Market.sol

```diff
@@ -0,0 +1,593 @@
+
+    function isContainerForReportingToken(ReportingToken _shadyToken) public constant returns (bool) {
+        if (_shadyToken.getTypeName() != "ReportingToken") {
+            return false;
+        }
+        bytes32 _shadyId = _shadyToken.getPayoutDistributionHash();
+        ReportingToken _reportingToken = reportingTokens[_shadyId];
+        return _reportingToken == _shadyToken;
+    }
+
+    function isContainerForShareToken(IShareToken _shadyShareToken) public constant returns (bool) {
+        if (_shadyShareToken.getTypeName() != "ShareToken") {
+            return false;
+        }
+        return getShareToken(_shadyShareToken.getOutcome()) == _shadyShareToken;
+    }
+
+    function isContainerForDisputeBondToken(DisputeBondToken _shadyBondToken) public constant returns (bool) {
+        if (automatedReporterDisputeBondToken == _shadyBondToken) {
+            return true;
+        }
+        if (limitedReportersDisputeBondToken == _shadyBondToken) {
+            return true;
+        }
+        if (allReportersDisputeBondToken == _shadyBondToken) {
+            return true;
+        }
+        return false;
+    }
+
+    function canBeReportedOn() public constant returns (bool) {
+        // CONSIDER: should we check if migration is necessary here?
+        if (isFinalized()) {
+            return false;
+        }
+        if (!reportingWindow.isReportingActive()) {
+            return false;
+        }
+        return true;
+    }
+
+    function needsMigration() public constant returns (bool) {
+        if (isFinalized()) {
+            return false;
+        }
+        Market _forkingMarket = getBranch().getForkingMarket();
+        if (address(_forkingMarket) == NULL_ADDRESS) {
+            return false;
+        }
+        if (_forkingMarket == this) {
+            return false;
+        }
+        if (block.timestamp < endTime) {
+            return false;
+        }
+        if (automatedReporterAddress != NULL_ADDRESS && block.timestamp < getAutomatedReportDueTimestamp()) {
+            return false;
+        }
+        if (automatedReportReceived && block.timestamp < getAutomatedReportDisputeDueTimestamp()) {
+            return false;
+        }
+        if (automatedReportReceived && address(automatedReporterDisputeBondToken) == NULL_ADDRESS) {
+            return false;
+        }
+        return true;
+    }
+
+    function getAutomatedReportDueTimestamp() public constant returns (uint256) {
+        return endTime + AUTOMATED_REPORTING_DURATION_SECONDS;
+    }
+
+    function getAutomatedReportDisputeDueTimestamp() public constant returns (uint256) {
+        return getAutomatedReportDueTimestamp() + AUTOMATED_REPORTING_DISPUTE_DURATION_SECONDS;
+    }
+
+    function getReportingState() public constant returns (ReportingState) {
+        if (isFinalized()) {
+            return ReportingState.FINALIZED;
+        }
+
+        if (needsMigration()) {
+            return ReportingState.AWAITING_MIGRATION;
+        }
```

Comment:
I wonder if we should just integrate `needsMigration` into this method instead of having it be a separate method?  More specifically, I wonder if up in `migrateThroughOneFork` we should call `getReportingState` and compare it against `AWAITING_MIGRATION` instead of calling `needsMigration` directly.  This would make `needsMigration` just an implementation detail of `getReportingState` rather than something that needs to stand on its own.